DuckDuckGo/TabSwitcherViewController.swift

```diff
@@ -72,6 +73,51 @@ class TabSwitcherViewController: UIViewController {
         titleView.text = count == 0 ? UserText.tabSwitcherTitleNoTabs : UserText.tabSwitcherTitleHasTabs
     }
     
+    fileprivate func displayBookmarkStatusToast(_ savedBookmarks: [Int], _ alreadySavedBookmarks: [Int]) {
+        if (tabsModel.tabs.count == savedBookmarks.count) || (tabsModel.tabs.count == savedBookmarks.count + alreadySavedBookmarks.count) {
+            self.view.showBottomToast(UserText.bookmarkAllTabsSaved)
+        } else {
+            let failedSaveCount = tabsModel.tabs.count - savedBookmarks.count - alreadySavedBookmarks.count
+            let statusMessage = String.localizedStringWithFormat(UserText.bookmarkAllTabsFailedToSave, failedSaveCount, tabsModel.tabs.count)
+            self.view.showBottomToast(statusMessage)
+        }
+    }
+    
+    fileprivate func bookmarkAll(_ openTabs: [Tab] ) -> ([Int], [Int]) {
+        
+        let bookmarksManager = BookmarksManager()
+        // here tab index are stored in arrays, in case we need to close bookmarked tabs
+        var savedBookmarks: [Int] = []
+        var alreadySavedBookmarks: [Int] = []
+        
+        for (index, aTab) in openTabs.enumerated() {
+            if let link = aTab.link {
+                if bookmarksManager.contains(url: link.url) {
+                    alreadySavedBookmarks.append(index)
+                    Logger.log(text: "\(link.url) already bookmarked")
+                } else {
+                    bookmarksManager.save(bookmark: link)
+                    savedBookmarks.append(index)
+                }
+            } else {
+                Logger.log(text: "no valid link found for tab \(aTab)")
+            }
+        }
+        return (savedBookmarks: savedBookmarks, alreadySavedBookmarks: alreadySavedBookmarks)
+    }
+    
+    @IBAction func onBookmarkAllOpenTabsPressed(_ sender: UIButton) {
+        
+        guard tabsModel.tabs.count > 0 else {
+            self.view.showBottomToast(UserText.bookmarkAllTabsNotFound)
+            Logger.log(text: "No open tabs found to bookmark")
```

Comment:
We probably don't want to log this as this is a regular scenario.

Reply:
sure..